File: python/gridwell/_gt_emitter.py
"""Convert a Great Tables GT object to gridwell IR JSON."""

from __future__ import annotations

import json
from typing import Any

import pandas as pd


def gt_to_ir(gt_obj: Any) -> str:
    """Convert a Great Tables GT object to gridwell IR JSON.

    Parameters
    ----------
    gt_obj : great_tables.GT
        A Great Tables GT object.

    Returns
    -------
    str
        JSON string in gridwell IR format.
    """
    return json.dumps(_gt_to_dict(gt_obj), indent=2)


def gt_to_dict(gt_obj: Any) -> dict:
    """Convert a Great Tables GT object to gridwell IR dict.

    Parameters
    ----------
    gt_obj : great_tables.GT
        A Great Tables GT object.

    Returns
    -------
    dict
        Dictionary in gridwell IR format.
    """
    return _gt_to_dict(gt_obj)


def _gt_to_dict(gt_obj: Any) -> dict:
    """Internal: build the IR dictionary from a GT object."""
    from great_tables._utils_render_html import cast_frame_to_string, replace_null_frame

    # Build the GT to resolve formats
    built = gt_obj._build_data("html")._render_formats("html")

    # Resolve cell text: formatted values override raw data
    str_orig = cast_frame_to_string(built._tbl_data)
    tbl_data = replace_null_frame(built._body.body, str_orig)

    # Column info
    visible_cols = built._boxhead._get_default_columns()
    stub_col_info = built._boxhead._get_stub_column()
    has_stub = stub_col_info is not None

    # All display columns (stub first if present, then data columns)
    all_cols = []
    if has_stub:
        all_cols.append(stub_col_info)
    all_cols.extend(visible_cols)

    n_cols = len(all_cols)
    stub_cols = 1 if has_stub else 0

    # Row grouping
    group_rows = built._stub.group_rows if hasattr(built._stub, "group_rows") else None
    has_groups = group_rows is not None and len(group_rows) > 0

    # Build config
    n_body_rows = len(tbl_data)
    config = _build_config(n_cols, n_body_rows, stub_cols, built)

    # Build column_spec
    column_spec = _build_column_spec(all_cols, has_stub)

    # Build header
    header = _build_header(built._heading)

    # Build thead
    thead = _build_thead(all_cols)

    # Build tbody (with row groups)
    tbody = _build_tbody(tbl_data, all_cols, built, has_stub, has_groups, group_rows)

    # Build footer
    footer = _build_footer(built._source_notes, built._footnotes)

    # Build styles
    styles = _build_styles(built._styles)

    return {
        "ir_version": "1.0",
        "config": config,
        "styles": styles,
        "header": header,
        "column_spec": column_spec,
        "table": {
            "thead": thead,
            "tbody": tbody,
        },
        "footer": footer,
        "extensions": {},
    }


def _build_config(n_cols: int, n_body_rows: int, stub_cols: int, built: Any) -> dict:
    """Build the config section."""
    return {
        "table_cols": n_cols,
        "header_rows": 1,
        "body_rows": n_body_rows,
        "stub_cols": stub_cols,
        "row_striping": False,
        "row_striping_include_stub": False,
        "row_striping_include_body": False,
        "column_labels_hidden": False,
        "table_width": None,
        "container_width": None,
        "container_height": None,
        "container_overflow": None,
        "locale": "en-US",
        "page_break_mode": "avoid",
        "aria_label": None,
        "aria_describedby": None,
        "summary": None,
    }


def _build_column_spec(all_cols: list, has_stub: bool) -> list[dict]:
    """Build column_spec array from ColInfo objects."""
    specs = []
    for i, col in enumerate(all_cols):
        is_stub_col = has_stub and i == 0
        specs.append({
            "id": col.var,
            "align": col.column_align or "left",
            "align_char": None,
            "width": col.column_width or "auto",
            "min_width": None,
            "max_width": None,
            "style_id": None,
            "hidden": False,
            "label": str(col.column_label) if col.column_label else col.var,
        })
    return specs


def _build_header(heading: Any) -> dict:
    """Build the header section from GT Heading."""
    title = None
    subtitle = None

    if heading.title:
        title = {"content": [{"type": "text", "value": str(heading.title)}]}
    if heading.subtitle:
        subtitle = {"content": [{"type": "text", "value": str(heading.subtitle)}]}

    return {
        "title": title,
        "subtitle": subtitle,
        "extra_lines": [],
        "preheader_content": None,
    }


def _build_thead(all_cols: list) -> dict:
    """Build the thead section with column labels."""
    cells = []
    for col in all_cols:
        label = str(col.column_label) if col.column_label else col.var
        cells.append({
            "content": [{"type": "text", "value": label}],
            "colspan": 1,
            "rowspan": 1,
            "style_id": None,
            "is_stub": False,
            "is_placeholder": False,
            "scope": "col",
            "sort_key": None,
            "data_type": None,
        })

    return {
        "rows": [{
            "role": "column_label",
            "style_id": None,
            "cells": cells,
        }],
    }


def _build_tbody(
    tbl_data: pd.DataFrame,
    all_cols: list,
    built: Any,
    has_stub: bool,
    has_groups: bool,
    group_rows: Any,
) -> list[dict]:
    """Build the tbody section with row groups."""
    if has_groups and group_rows:
        groups = []
        for grp in group_rows:
            rows = []
            for idx in grp.indices:
                row = _build_row(tbl_data, int(idx), all_cols, built, has_stub)
                rows.append(row)

            label = None
            if grp.group_label:
                label = {
                    "content": [{"type": "text", "value": str(grp.group_label)}],
                    "style_id": None,
                    "colspan": None,
                }

            groups.append({
                "group_id": grp.group_id,
                "label": label,
                "rows": rows,
                "summary_rows": [],
            })
        return groups
    else:
        # Single group with all rows
        rows = []
        for idx in range(len(tbl_data)):
            row = _build_row(tbl_data, idx, all_cols, built, has_stub)
            rows.append(row)
        return [{
            "group_id": None,
            "label": None,
            "rows": rows,
            "summary_rows": [],
        }]


def _build_row(
    tbl_data: pd.DataFrame,
    row_idx: int,
    all_cols: list,
    built: Any,
    has_stub: bool,
) -> dict:
    """Build a single row."""
    cells = []
    for i, col in enumerate(all_cols):
        is_stub = has_stub and i == 0
        value = tbl_data[col.var].iloc[row_idx]
        text = str(value) if not pd.isna(value) else ""

        # Determine data type from the original data
        orig_val = built._tbl_data[col.var].iloc[row_idx]
        data_type = _infer_data_type(orig_val, col.var, built._tbl_data.dtypes)

        cell = {
            "content": [{"type": "text", "value": text}],
            "colspan": 1,
            "rowspan": 1,
            "style_id": None,
            "is_stub": is_stub,
            "is_placeholder": False,
            "data_type": data_type,
            "sort_key": text.lower() if text else None,
        }

        # Add typed_value for non-stub cells
        if not is_stub and text:
            cell["typed_value"] = {
                "type": data_type or "string",
                "value": str(orig_val) if not pd.isna(orig_val) else text,
            }

        cells.append(cell)

    return {
        "role": None,
        "style_id": None,
        "cells": cells,
    }


def _infer_data_type(value: Any, col_name: str, dtypes: Any) -> str | None:
    """Infer the gridwell data_type from a pandas value/dtype."""
    if pd.isna(value):
        return None

    dtype = dtypes.get(col_name)
    if dtype is not None:
        dtype_str = str(dtype)
        if "float" in dtype_str:
            return "number"
        elif "int" in dtype_str:
            return "integer"
        elif "datetime" in dtype_str:
            return "datetime"
        elif "bool" in dtype_str:
            return "boolean"

    # Heuristic from value
    if isinstance(value, (int,)):
        return "integer"
    elif isinstance(value, (float,)):
        return "number"
    elif isinstance(value, bool):
        return "boolean"

    return "string"


def _build_footer(source_notes: list, footnotes: list) -> dict:
    """Build the footer section."""
    fn_list = []
    for i, fn in enumerate(footnotes):
        fn_text = str(fn) if not isinstance(fn, dict) else fn.get("text", str(fn))
        fn_list.append({
            "id": f"fn{i + 1}",
            "content": [{"type": "text", "value": fn_text}],
        })

    sn_list = []
    for sn in source_notes:
        sn_text = str(sn) if not isinstance(sn, dict) else sn.get("text", str(sn))
        sn_list.append({
            "content": [{"type": "text", "value": sn_text}],
        })

    return {
        "footnotes": fn_list,
        "source_notes": sn_list,
    }


def _build_styles(styles: list) -> dict:
    """Build the styles palette from GT StyleInfo entries."""
    defs: dict[str, dict] = {}

    for i, style_info in enumerate(styles):
        style_id = f"style_{i}"
        style_def: dict[str, Any] = {}

        # Extract style properties from the StyleInfo
        if hasattr(style_info, "styles"):
            for s in style_info.styles:
                if hasattr(s, "color") and s.color:
                    style_def["color"] = s.color
                if hasattr(s, "bgcolor") and s.bgcolor:
                    style_def["background_color"] = s.bgcolor
                if hasattr(s, "weight") and s.weight:
                    style_def["font_weight"] = s.weight
                if hasattr(s, "style") and s.style:
                    style_def["font_style"] = s.style
                if hasattr(s, "size") and s.size:
                    style_def["font_size"] = s.size
                if hasattr(s, "align") and s.align:
                    style_def["text_align"] = s.align

        if style_def:
            defs[style_id] = style_def

    return {
        "defs": defs,
        "compositions": {},
        "conditionals": [],
    }

File: r-package/R/gt_emitter.R
#' Convert a gt table object to gridwell IR JSON
#'
#' @param gt_obj A `gt_tbl` object from the gt package.
#' @return A JSON string in gridwell IR format.
#' @export
gt_to_ir <- function(gt_obj) {
    ir <- gt_to_ir_list(gt_obj)
    jsonlite::toJSON(ir, auto_unbox = TRUE, null = "null", pretty = TRUE)
}

#' Convert a gt table object to a gridwell IR list
#'
#' @param gt_obj A `gt_tbl` object from the gt package.
#' @return A list representing the gridwell IR.
#' @export
gt_to_ir_list <- function(gt_obj) {

    # Extract gt internals
    data <- gt_obj[["_data"]]
    boxhead <- gt_obj[["_boxhead"]]
    stub_df <- gt_obj[["_stub_df"]]
    heading <- gt_obj[["_heading"]]
    row_groups <- gt_obj[["_row_groups"]]
    source_notes <- gt_obj[["_source_notes"]]
    footnotes <- gt_obj[["_footnotes"]]

    # Determine visible columns (not stub, not row_group, not hidden)
    is_default <- boxhead$type == "default"
    is_stub <- boxhead$type == "stub"
    has_stub <- any(is_stub)

    visible_cols <- boxhead$var[is_default]
    stub_var <- if (has_stub) boxhead$var[is_stub][1] else NULL

    # All display columns
    all_vars <- if (has_stub) c(stub_var, visible_cols) else visible_cols
    n_cols <- length(all_vars)
    stub_cols <- if (has_stub) 1L else 0L

    has_groups <- length(row_groups) > 0

    # Build config
    config <- list(
        table_cols = n_cols,
        header_rows = 1L,
        body_rows = nrow(data),
        stub_cols = stub_cols,
        row_striping = FALSE,
        row_striping_include_stub = FALSE,
        row_striping_include_body = FALSE,
        column_labels_hidden = FALSE,
        table_width = NULL,
        container_width = NULL,
        container_height = NULL,
        container_overflow = NULL,
        locale = "en-US",
        page_break_mode = "avoid",
        aria_label = NULL,
        aria_describedby = NULL,
        summary = NULL
    )

    # Build column_spec
    column_spec <- lapply(all_vars, function(var) {
        idx <- which(boxhead$var == var)
        align <- boxhead$column_align[idx]
        width <- boxhead$column_width[[idx]]
        label <- boxhead$column_label[[idx]]
        if (is.null(label)) label <- var

        list(
            id = var,
            align = align,
            align_char = NULL,
            width = if (is.null(width)) "auto" else as.character(width),
            min_width = NULL,
            max_width = NULL,
            style_id = NULL,
            hidden = FALSE,
            label = as.character(label)
        )
    })

    # Build header
    header <- list(
        title = NULL,
        subtitle = NULL,
        extra_lines = list(),
        preheader_content = NULL
    )
    if (!is.null(heading$title) && nzchar(heading$title)) {
        header$title <- list(
            content = list(list(type = "text", value = heading$title))
        )
    }
    if (!is.null(heading$subtitle) && nzchar(heading$subtitle)) {
        header$subtitle <- list(
            content = list(list(type = "text", value = heading$subtitle))
        )
    }

    # Build thead
    thead_cells <- lapply(all_vars, function(var) {
        idx <- which(boxhead$var == var)
        label <- boxhead$column_label[[idx]]
        if (is.null(label)) label <- var

        list(
            content = list(list(type = "text", value = as.character(label))),
            colspan = 1L,
            rowspan = 1L,
            style_id = NULL,
            is_stub = FALSE,
            is_placeholder = FALSE,
            scope = "col",
            sort_key = NULL,
            data_type = NULL
        )
    })

    thead <- list(
        rows = list(list(
            role = "column_label",
            style_id = NULL,
            cells = thead_cells
        ))
    )

    # Build tbody
    tbody <- .build_tbody_r(data, all_vars, stub_df, has_stub, has_groups,
                            row_groups, stub_var)

    # Build footer
    footer <- list(
        footnotes = list(),
        source_notes = list()
    )
    if (length(source_notes) > 0) {
        footer$source_notes <- lapply(source_notes, function(sn) {
            text <- if (is.list(sn)) as.character(sn[[1]]) else as.character(sn)
            list(content = list(list(type = "text", value = text)))
        })
    }

    # Build styles (empty for now - gt styles are complex)
    styles <- list(
        defs = setNames(list(), character(0)),
        compositions = setNames(list(), character(0)),
        conditionals = list()
    )

    list(
        ir_version = "1.0",
        config = config,
        styles = styles,
        header = header,
        column_spec = column_spec,
        table = list(thead = thead, tbody = tbody),
        footer = footer,
        extensions = setNames(list(), character(0))
    )
}

.build_tbody_r <- function(data, all_vars, stub_df, has_stub, has_groups,
                           row_groups, stub_var) {
    if (has_groups && length(row_groups) > 0) {
        groups <- lapply(row_groups, function(grp_id) {
            row_indices <- which(stub_df$group_id == grp_id)
            rows <- lapply(row_indices, function(i) {
                .build_row_r(data, i, all_vars, has_stub, stub_var)
            })

            # Group label
            grp_labels <- stub_df$group_label[row_indices]
            grp_label_text <- as.character(grp_labels[[1]])

            label <- list(
                content = list(list(type = "text", value = grp_label_text)),
                style_id = NULL,
                colspan = NULL
            )

            list(
                group_id = grp_id,
                label = label,
                rows = rows,
                summary_rows = list()
            )
        })
        return(groups)
    } else {
        rows <- lapply(seq_len(nrow(data)), function(i) {
            .build_row_r(data, i, all_vars, has_stub, stub_var)
        })
        return(list(list(
            group_id = NULL,
            label = NULL,
            rows = rows,
            summary_rows = list()
        )))
    }
}

.build_row_r <- function(data, row_idx, all_vars, has_stub, stub_var) {
    cells <- lapply(seq_along(all_vars), function(col_idx) {
        var <- all_vars[col_idx]
        is_stub_cell <- has_stub && col_idx == 1
        value <- data[[var]][row_idx]
        text <- if (is.na(value)) "" else as.character(value)

        data_type <- .infer_dtype_r(value)

        cell <- list(
            content = list(list(type = "text", value = text)),
            colspan = 1L,
            rowspan = 1L,
            style_id = NULL,
            is_stub = is_stub_cell,
            is_placeholder = FALSE,
            data_type = data_type,
            sort_key = if (nzchar(text)) tolower(text) else NULL
        )

        if (!is_stub_cell && nzchar(text)) {
            cell$typed_value <- list(
                type = if (is.null(data_type)) "string" else data_type,
                value = as.character(value)
            )
        }

        cell
    })

    list(
        role = NULL,
        style_id = NULL,
        cells = cells
    )
}

.infer_dtype_r <- function(value) {
    if (is.na(value)) return(NULL)
    if (is.numeric(value)) {
        if (is.integer(value)) return("integer")
        return("number")
    }
    if (is.logical(value)) return("boolean")
    if (inherits(value, "Date")) return("date")
    if (inherits(value, "POSIXt")) return("datetime")
    "string"
}
